From f0d32b9fd9d5c3e15aaa765b388b4ea87e7cfb4c Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Mon, 22 Nov 2010 13:24:13 +0000
Patch-mainline: 2.6.38
References: fate#310916
Git-commit: ab5793ad3ae11a5cbe2194b449e5fdd80b19f14f
Subject: [PATCH 1898/2588] drm/i915: Tweak on-error bbaddr parsing for
 clarity

Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
(cherry picked from commit ab5793ad3ae11a5cbe2194b449e5fdd80b19f14f)

Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 drivers/gpu/drm/i915/i915_irq.c |   19 ++++++++-----------
 1 files changed, 8 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index be54a86..89029e6 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -526,26 +526,23 @@ i915_ringbuffer_last_batch(struct drm_device *dev,
 	/* Locate the current position in the ringbuffer and walk back
 	 * to find the most recently dispatched batch buffer.
 	 */
-	bbaddr = 0;
 	head = I915_READ_HEAD(ring) & HEAD_ADDR;
-	val = (u32 *)(ring->virtual_start + head);
 
+	val = (u32 *)(ring->virtual_start + head);
 	while (--val >= (u32 *)ring->virtual_start) {
 		bbaddr = i915_get_bbaddr(dev, val);
 		if (bbaddr)
-			break;
+			return bbaddr;
 	}
 
-	if (bbaddr == 0) {
-		val = (u32 *)(ring->virtual_start + ring->size);
-		while (--val >= (u32 *)ring->virtual_start) {
-			bbaddr = i915_get_bbaddr(dev, val);
-			if (bbaddr)
-				break;
-		}
+	val = (u32 *)(ring->virtual_start + ring->size);
+	while (--val >= (u32 *)ring->virtual_start) {
+		bbaddr = i915_get_bbaddr(dev, val);
+		if (bbaddr)
+			return bbaddr;
 	}
 
-	return bbaddr;
+	return 0;
 }
 
 static u32 capture_bo_list(struct drm_i915_error_buffer *err,
-- 
1.7.6

